sticker2/src/graph/mst.rs

```diff
@@ -0,0 +1,737 @@
+
+    fn assert_tree(parents: &[Option<usize>], root: usize) {
+        for (vertex, &parent) in parents.iter().enumerate() {
+            if vertex == root {
+                assert_eq!(
+                    parent, None,
+                    "Root vertex {} has a parent in graph {:?}",
+                    root, parents
+                )
+            } else {
+                assert!(
+                    parent.is_some(),
+                    "Non-root vertex {} does not have a parent in the graph {:?}",
+                    vertex,
+                    parents
+                )
+            }
+        }
+
+        let cycle = find_cycle(parents);
+        assert_eq!(
+            find_cycle(parents),
+            None,
+            "Graph {:?} contains a cycle: {:?}",
+            parents,
+            cycle.unwrap()
+        );
+    }
+
+    #[test]
+    pub fn finds_max_parents() {
+        let distances = Array::range(0f32, 25f32, 1f32).into_shape((5, 5)).unwrap();
+        let max_parents = find_max_parents(distances.view(), 0, &[true; 5]);
+        assert_eq!(max_parents, vec![None, Some(4), Some(4), Some(4), Some(3)]);
+    }
+
+    #[test]
+    pub fn finds_max_parents_with_inactive_vertices() {
+        let distances = Array::range(0f32, 25f32, 1f32).into_shape((5, 5)).unwrap();
+        let max_parents = find_max_parents(distances.view(), 0, &[true, false, true, false, true]);
+        assert_eq!(max_parents, vec![None, None, Some(4), None, Some(2)]);
+    }
+
+    #[test]
+    pub fn finds_trees_in_random_graphs() {
+        // We should probably use quickcheck or proptest for this, but
+        // then I have to figure out how to do proper shrinkage, since
+        // we require square matrices. For now, we are just happy to now
+        // if we produce proper trees.
+
+        const NUM_TEST_ITERATIONS: usize = 1000;
+
+        let mut rng = XorShiftRng::seed_from_u64(42);
+        for _ in 0..NUM_TEST_ITERATIONS {
+            let scores = Array::random_using((10, 10), Uniform::new(0f32, 1f32), &mut rng);
+            let mst = chu_liu_edmonds(scores.view(), 0);
+            assert_tree(&mst, 0);
+        }
+    }
+
+    #[test]
+    pub fn finds_cycle() {
+        // No cycle.
+        assert_eq!(
+            find_cycle(&[None, Some(0), Some(1), Some(2), Some(3)]),
+            None,
+        );
+
+        // No cycle.
+        assert_eq!(
+            find_cycle(&[None, Some(0), Some(0), Some(0), Some(0)]),
+            None,
+        );
+
+        // Short cycle: 3 -> 4 -> 3
+        assert_eq!(
+            find_cycle(&[None, Some(4), Some(4), Some(4), Some(3)]),
+            Some(vec![3, 4])
+        );
+
+        // Long cycle: 1 -> 2 -> 3 -> 4 -> 1
+        assert_eq!(
+            find_cycle(&[None, Some(4), Some(1), Some(2), Some(3)]),
+            Some(vec![2, 3, 4, 1])
+        );
+
+        // Self-cycle
+        assert_eq!(find_cycle(&[Some(0)]), Some(vec![0]));
+    }
```

Comment:
Add test with more than one cycle to make sure recursion works as expected.

Reply:
Note that `find_cycle` does not recurse, it just finds an arbitrary cycle if it exists.

The recursion case for Chu-Liu-Edmonds is tested in `correctly_decodes_toy` matrices, since the 4x4 matrix wil result in another cycle when the first cycle is resolved. Also the `finds_trees_in_random_graphs` will typically contain cycles given the size of the matrix and randomness.

Reply:
I see 😊